src/legacy-appcache-behavior-import.js

```diff
@@ -0,0 +1,422 @@
+        });
+      }));
+    });
+  }
+
+  /**
+   * Fulfills with an array of all the hash ids that correspond to outdated
+   * manifest versions.
+   *
+   * @returns {Promise.<Array.<String>>}
+   */
+  function getHashesOfOlderVersions() {
+    return idb.getAll(constants.STORES.MANIFEST_URL_TO_CONTENTS).then(manifests => {
+      return manifests.map(versions => {
+        // versions.slice(0, -1) will give all the versions other than the
+        // last, or [] if there's aren't any older versions.
+        return versions.slice(0, -1).map(version => version.hash);
+      }).reduce((prev, curr) => {
+        // Flatten the array-of-arrays into an array.
+        return prev.concat(curr);
+      }, []);
+    });
+  }
+
+  /**
+   * Does the following:
+   * 1. Gets a list of all client ids associated with this service worker.
+   * 2. Calls cleanupClientIdAndHash() to remove the out of date client id
+   *    to hash associations.
+   * 3. Calls getHashesOfOlderVersions() to get a list of all the hashes
+   *    that correspond to out-of-date manifest versions.
+   * 4. If there's a match between an out of date hash and a hash that is no
+   *    longer being used by a client, then it deletes the corresponding cache.
+   *
+   * @returns {Promise.<T>}
+   */
+  function cleanupOldCaches() {
+    self.clients.matchAll().then(clients => {
+      return clients.map(client => client.id);
+    }).then(idsOfActiveClients => {
+      return cleanupClientIdAndHash(idsOfActiveClients);
+    }).then(hashesNotInUse => {
+      return getHashesOfOlderVersions().then(hashesOfOlderVersions => {
+        return hashesOfOlderVersions.filter(hashOfOlderVersion => {
+          return hashesNotInUse.includes(hashOfOlderVersion);
+        });
+      });
+    }).then(idsToDelete => {
+      log('deleting cache ids', idsToDelete);
+      return Promise.all(idsToDelete.map(cacheId => caches.delete(cacheId)));
+    });
+
+    // TODO: Delete the entry in the array stored in MANIFEST_URL_TO_CONTENT.
+  }
+
+  /**
+   * A wrapper on top of appCacheBehaviorForEvent() that handles rejections with
+   * a default of fetch().
+   *
+   * @param {FetchEvent} event
+   * @returns {Promise.<Response>}
+   */
+  global.legacyAppCacheBehavior = event => {
+    log('client id is', event.clientId);
+    return appCacheBehaviorForEvent(event).then(response => {
+      // If this is a navigation, clean up unused caches that correspond to old
+      // AppCache manifest versions which are no longer associated with an
+      // active client. This will be done asynchronously, and won't block the
+      // response from being returned to the onfetch handler.
+      if (event.request.mode === 'navigate') {
+        cleanupOldCaches();
+      }
+
+      return response;
+    }).catch(error => {
+      console.warn('No AppCache behavior for', event.request.url, error);
+      // TODO: Is it sensible to use fetch() here as a fallback?
```

Comment:
IMHO, this does not make sense.

You handled all the possible outcomes for a request previously: if there is not manifest, you handled it at [`noManifestBehavior()`](https://github.com/jeffposnick/appcache2serviceworker/pull/1/files?diff=unified#diff-cbaca1f6f5d61913849758c592315c1aR225); if there is manifest but there is no behavior, you handled it at [`appCacheLogic()`](https://github.com/jeffposnick/appcache2serviceworker/pull/1/files?diff=unified#diff-cbaca1f6f5d61913849758c592315c1aR181) as well. So this `catch()` is only for error handling. If you leave this, you silence the errors in favor of `fetch()` so only further network errors leave your API.

This should be handled by the [consumer](https://github.com/jeffposnick/appcache2serviceworker/pull/1/files?diff=unified#diff-b1e184332a58c2c8dadd8c2a62809114R4):
```js
self.addEventListener('fetch', event => {
  event.respondWith(legacyAppCacheBehavior(event).catch(error => fetch(event.request)));
});
```